tests/unit/s2n_extensions_server_key_share_select_test.c

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+#include "tls/s2n_tls.h"
+#include "tls/s2n_tls13.h"
+#include "tls/extensions/s2n_server_key_share.h"
+
+int main(int argc, char **argv)
+{
+    struct s2n_connection *server_conn;
+
+    BEGIN_TEST();
+    
+    EXPECT_SUCCESS(s2n_enable_tls13());
+
+    /* These tests check the server's logic when selecting a keyshare and supporting group. */
+
+    {
+        /* If client and server have no mutually supported groups and no mutually supported
+         * keyshares, a Hello Retry Request is not sent. 
+         */
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_NULL(server_conn->secure.server_ecc_evp_params.negotiated_curve);
+        for (int i = 0; i < S2N_ECC_EVP_SUPPORTED_CURVES_COUNT; i++) {
+            EXPECT_NULL(server_conn->secure.client_ecc_evp_params[i].evp_pkey);
+            EXPECT_NULL(server_conn->secure.client_ecc_evp_params[i].negotiated_curve);
+        }
+
+        EXPECT_SUCCESS(s2n_extensions_server_key_share_select(server_conn));
+
+        EXPECT_NULL(server_conn->secure.server_ecc_evp_params.negotiated_curve);
+        EXPECT_FALSE(s2n_is_hello_retry_req(server_conn));
```

Comment:
`s2n_is_hello_retry_req` will be called from the client. I'm using a `s2n_is_hello_retry_required()` function which may be better here.